### PR TITLE
Residential: connect to ResStock

### PR DESCRIPTION
**Pull request by joseph-robertson**
This PR makes several changes/updates. They are summarized into the following items:
* **(A)** Replace `hpxml-measures` with `residential-measures`
* **(B)** Refactor Baseline.rb's residential paths
* **(C)** Write new `residential_samples` path
* **(D)** Write new `BuildResidentialModel` unit tests

Details are given below.

**(A)** Replace `example_project/resources/hpxml-measures` with `example_project/resources/residential-measures`. The latter has `hpxml-measures` in it. A new `update_residential` rake task is created for git cloning (as opposed to git subtree) the resstock repository with a depth of 1 and renaming it to `residential-measures`. The rake task also removes a bunch of files and folders that UO doesn't need.

**(B)** The set of residential paths are now:
* Argument values assigned using OpenStudio-HPXML defaults, or
* Defaults + Argument values assigned using `residential_template` (customizable lookup files), or
* Defaults + Argument values assigned using `residential_samples` (ResStock parameter/options)

**(C)** To the Baseline.rb mapper, add optional `residential_samples` method (similar to the optional `residential_template` method).

The `residential_samples` path connects to a user-specified ResStock buildstock CSV file:
* if **resstock_buildstock_csv_path** is populated in the GeoJSON file, finds/sets a `resstock_building_id` for URBANopt's `BuildResidentialModel` measure
* if **uo_buildstock_mapping_csv_path** is populated in the GeoJSON file, uses the `resstock_building_id` mapped by feature ID for URBANopt's `BuildResidentialModel` measure

Both **resstock_buildstock_csv_path** and **uo_buildstock_mapping_csv_path** CSV files may be a subset of all ResStock parameters. Any arguments not assigned because of missing parameters will be filled in with defaults.

Copy some code from ResStock's `BuildExistingModel` measure into `BuildResidentialModel` measure. (It probably doesn't make sense to call ResStock's `BuildExistingModel` directly because it currently creates HPXML files with only a single dwelling unit. Even if it could create multiple dwelling units, they'd all be identical?)

**(D)** Add new `test_residential_samplesXXX` tests to the `test_build_residential_model.rb` test file that checks combinations of buildstock CSV file paths and resstock building IDs.

--------------

Checklist / Questions:
- [x] make a better distinction between the "Feature ID" and the "ResStock Building ID" (`building_id` -> `resstock_building_id`)
- [x] resstock subtree considerations
  * think about whether we need to remove some resstock repo files (it's pretty huge)
    * ideally, yes, we wouldn't want any resstock files that we don't actually need
  * is it confusing to replace `hpxml-measures` with `resstock/resources/hpxml-measures`? seems redundant to store both.
    * it's confusing to have folder name "resstock" within UO; perhaps we could change the folder name to something like "residential" or "residential-measures"?
  * really we only want "hpxml-measures", "options_lookup.tsv", "ResStockArguments", and maybe a few other resource files
- [x] address all FIXMEs and TODOs
- [ ] design/implement the geojson schema changes https://github.com/urbanopt/urbanopt-geojson-gem/pull/270
- [ ] add a new sample feature to `example_project_combined.json` that demonstrates the new `residential_samples` path?
- [ ] should each dwelling unit of a building be identical when using the ResStock path? or should some options in the lookup be ignored (i.e., foundation/attic type, unit vertical/horizontal location relative the building)?
  - Example 1: for a MF building, we determine foundation/attic type and orientation for each dwelling unit.
  - Example 2: weather station epw is assigned in the mapper.
  * initial thought is that we should follow UO's logic for determining unit positions; this gives variability that we otherwise wouldn't have
  * follow-up thought here is that we shouldn't; e.g., duct location can conflict with "new" foundation type
- [ ] think about how to throw a warning in case the user-provided buildstock CSV file does not sync with the checkout of resstock? maybe it's enough to just call `test/check_buildstock.rb` here?
- [x] why doesn't feature.osw get written to the test run folder?

### CODE REVIEW

File: example_project/measures/BuildResidentialModel/measure.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@
 # http://nrel.github.io/OpenStudio-user-documentation/measures/measure_writing_guide/
 
 require 'openstudio'
-resources_path = File.absolute_path(File.join(File.dirname(__FILE__), '../../resources/hpxml-measures/HPXMLtoOpenStudio/resources'))
+resources_path = File.absolute_path(File.join(File.dirname(__FILE__), '../../resources/resstock/resources/hpxml-measures/HPXMLtoOpenStudio/resources'))
```
Comment (joseph-robertson) [inline example_project/measures/BuildResidentialModel/measure.rb:10]: Yes, we're now using files from a resource folder within a resource folder within a resource folder...

File: example_project/measures/BuildResidentialModel/measure.rb [hunk 2]
```diff
@@ -72,7 +72,12 @@ def arguments(model)
     arg.setDescription('The name of the folder containing a custom HPXML file, relative to the xml_building folder.')
     args << arg
 
-    measures_dir = File.absolute_path(File.join(File.dirname(__FILE__), '../../resources/hpxml-measures'))
+    arg = OpenStudio::Measure::OSArgument.makeIntegerArgument('building_id', false)
```
Comment (joseph-robertson) [inline example_project/measures/BuildResidentialModel/measure.rb:75]: If we copy the user-specified buildstock CSV file to a static location prior to this measure, I don't think we'd need to expose a buildstock CSV filepath argument here.

File: example_project/measures/BuildResidentialModel/measure.rb [hunk 3]
```diff
@@ -108,10 +113,85 @@ def run(model, runner, user_arguments)
...
+      if !apply_measures(measures_dir, { 'ResStockArguments' => measures['ResStockArguments'] }, resstock_arguments_runner, model, true, 'OpenStudio::Measure::ModelMeasure', nil)
+        return false
+      end
+
+      resstock_arguments_runner.result.stepValues.each do |step_value|
+        value = get_value_from_workflow_step_value(step_value)
+        next if value == ''
+        next if step_value.name == 'weather_station_epw_filepath' # we already assign this in the Baseline.rb mapper
+
+        args[step_value.name.to_sym] = value # FIXME: we'd probably want to check that lookup assignments don't conflict with geojson assignments?
```
Comment (joseph-robertson) [inline example_project/measures/BuildResidentialModel/measure.rb:192]: Yes, a lot of this (and code above) is copied from BuildExistingModel. See the PR description for an explanation.

File: example_project/mappers/Baseline.rb [hunk 4]
```diff
@@ -1497,6 +1222,273 @@ def time_mapping(time)
...
+        resstock_dir = File.absolute_path(File.join(File.dirname(__FILE__), '../resources/resstock'))
+        lib_dir = File.join(resstock_dir, 'lib')
+        resources_dir = File.join(resstock_dir, 'resources')
+        housing_characteristics_dir = File.join(resstock_dir, 'project_national/housing_characteristics')
+
+        FileUtils.rm_rf(lib_dir)
+        Dir.mkdir(lib_dir)
+        FileUtils.cp_r(resources_dir, lib_dir)
+        FileUtils.cp_r(housing_characteristics_dir, lib_dir)
+        FileUtils.cp(buildstock_csv_path, File.join(housing_characteristics_dir, 'buildstock.csv'))
```
Comment (joseph-robertson) [inline example_project/mappers/Baseline.rb:1490]: I'm not sure if ultimately we'd _need_ all this `lib` folder business.

**Comment by joseph-robertson** [comment]
Questions after running Feature ID = 18, which returns Building ID = 1654.

They have different:
* foundation/attic types
* HVAC system types
* heating fuel types

Should we be matching on these parameters/properties, or does this constrain the search too much?

Currently if no match is found, we error out. Is there a better / more flexible approach here? Should we iteratively remove a parameter, re-do the search, etc etc?

File: example_project/measures/BuildResidentialModel/measure.rb (hunk 2 above)
Comment (joseph-robertson) [reply example_project/measures/BuildResidentialModel/measure.rb]: Decided to expose a buildstock CSV filepath argument. Then we can remove all the lib folder business.

File: example_project/mappers/Baseline.rb (hunk 4 above)
Comment (joseph-robertson) [reply example_project/mappers/Baseline.rb]: We don't.

**Comment by joseph-robertson** [comment]
> Questions after running Feature ID = 18, which returns Building ID = 1654.
> 
> They have different:
> 
> * foundation/attic types
> * HVAC system types
> * heating fuel types
> 
> Should we be matching on these parameters/properties, or does this constrain the search too much?
> 
> Currently if no match is found, we error out. Is there a better / more flexible approach here? Should we iteratively remove a parameter, re-do the search, etc etc?

Seems like we may want to update the GeoJSON schema such that when `resstock_buildstock_csv_path` or `uo_buildstock_mapping_csv_path` is present, other required fields like `foundation_type` and `attic_type` are not required. Otherwise you can run into trouble:
* you'll either override the feature-specified foundation/attic type (and corresponding insulation) based on the selected resstock sample, or
* you'll ignore the foundation/attic type override but also need to ignore corresponding sampled insulation

**Comment by joseph-robertson** [comment]
Some meeting notes:
* general approach will be to respect what's in either buildstock csv, with the exception of `geometry_unit_cfa`
* `resstock_buildstock_csv_path`
  * should come with a warning that random could lead to issues
  * more properties in geojson specified, less samples matched
  * also depends on size of buildstock.csv
* `uo_buildstock_mapping_csv_path`
  * should emphasize that this is an improvement or more innovative where the user has the ability to specify the appropriate set of characteristics
  * the responsibility is on the user to ensure they supply contradictory parameter/options

**Comment by joseph-robertson** [comment]
Meeting notes:
* we should have instructions regarding (a) how to get a buildstock.csv, and (b) other details (where to put it, etc etc)
* need to double-check whether "random" is seeded or not, so we don't get different results in a re-run
* what is the buildstock.csv "schema"? point to resstock RTD? to options_lookup? to example buildstock.csv?

**Comment by vtnate** [comment]
@joseph-robertson Does this PR resolve https://github.com/urbanopt/urbanopt-example-geojson-project/issues/192?

**Comment by vtnate** [comment]
@joseph-robertson Thanks for sorting out that question of #192 on the other thread. One more question: Does the residential measure code change very often? Is it worth putting the rake test task in CI and running weekly? I'm running it locally now and it takes quite a while.

**Review by kflemin — APPROVED** [review]

**Pull request merged**